Improve stocktake detail view speed

mark-prins opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #2904
And [38](https://github.com/msupply-foundation/open-msupply-internal/issues/38) and [41](https://github.com/msupply-foundation/open-msupply-internal/issues/41)

# 👩🏻‍💻 What does this PR do? 
 <!-- Explain the changes you made, and why they're needed. Add a screenshot if you've made any UI changes!  -->
The primary change here is to pagination the detail view stocktake_lines.

Other minor changes: tweaking the hook usage and improving UI responsiveness.
I've refactored into multiple components, to improve the modal loading and reduce re-renders. Have changed the sort and filter to be server-side now, and updated to use location code rather than name.

I refactored the hook usage - as the stocktake query was being called many times - so I tried to consolidate. Called the query once and passed around the resulting data object. This didn't significantly improve performance, so that is mostly reverted now. We did only call the query once in the original version, so that is still fine.

Two small UI tweaks - the `DetailTabs` had been altered to retain URL values, and in doing so, this meant that the first tab was alway loaded, even if the URL was for a different tab. In the stocktake detail page, this is important, as the first tab is very slow. I've changed how this operates now. Secondly, the 'group by item' switch was lagging terribly - as it rendered the state change only when the table was fully rendered. I've divorced the two states, so that it responds immediately and the table update follows after.

# 🧪 How has/should this change been tested? 
<!-- Explain how to setup for testing here if it is not already obvious, and how you've tested this PR. -->
Best to use the [Cooks database](https://drive.google.com/file/d/1fdoYKhSzlEZz645DJGiZvB0B05JJAGkB/view?usp=drive_link) but you can replicate with any database that has a lot of items. Create a stocktake with a few hundred lines and test on desktop. For me, the page takes 1.5s on average. It's a lot worse on android.

## 💌 Any notes for the reviewer?
<!-- eg. Do you have any specific questions for the reviewer? Is there a high risk/complicated change they should focus on? If there are any general areas of the codebase your changes might have have touched or could cause side effects to, mention them here.-->
We should check the `DetailTabs` change is benign. Other changes are isolated to the stocktake detail view. 
Test the filter and sorting too please. 
Try to edit a stocktake line, and to add a line.

## 📃 Documentation
<!-- Note down any areas which require documentation updates -->
_No user facing changes_

mark-prins commented on client/packages/common/src/hooks/useEditModal/useEditModal.ts:
```diff
@@ -20,17 +20,20 @@ export const useEditModal = <T>(): EditModalState<T> => {
-    setMode(entity ? ModalMode.Update : ModalMode.Create);
-    modalControl.toggleOn();
-  }, []);
+  const onOpen = useCallback(
```
just a linter request. usually they are bogus, but I think this is valid.

mark-prins commented on client/packages/common/src/types/schema.ts:
```diff
@@ -4822,7 +4823,7 @@ export enum StocktakeLineSortFieldInput {
   ItemCode = 'itemCode',
   ItemName = 'itemName',
-  LocationName = 'locationName',
+  LocationCode = 'locationCode',
```
this was a general request, that we use location code not name. took the opportunity to update.

mark-prins replied:
Louisa was the instigator and she was backed up by others who said that code is used in preference to name. the name is just too long to show in tables.
I took the YAGNI principle - we aren't planning to use name as a column or sort option, so I haven't implemented.

mark-prins commented on client/packages/common/src/ui/components/navigation/Tabs/DetailTabs.tsx:
```diff
@@ -29,9 +29,14 @@ export const DetailTabs: FC<DetailTabsProps> = ({
   const { urlQuery, updateQuery } = useUrlQuery();
-  const [currentTab, setCurrentTab] = useState<string>(tabs[0]?.value ?? '');
   const t = useTranslation();
+  const currentUrlTab = urlQuery['tab'] as string | undefined;
```
relying on the `useEffect` resulted in the first tab always loading. For stocktake detail view, trying to view the log was super slow because of this.

mark-prins commented on client/packages/inventory/src/Stocktake/DetailView/AppBarButtons.tsx:
```diff
@@ -26,11 +27,11 @@ interface AppBarButtonProps {
-  const { data } = useStocktake.document.get();
   const { print, isPrinting } = useReport.utils.print();
+  const { data } = useStocktake.document.get();
+  const isDisabled = !data || isStocktakeDisabled(data);
```
the `utils.isDisabled` hook is requesting the document too, so I've consolidated

mark-prins commented on client/packages/inventory/src/Stocktake/DetailView/ContentArea/ContentArea.tsx:
```diff
@@ -89,13 +89,16 @@ export const ContentArea: FC<ContentAreaProps> = ({
   const t = useTranslation('inventory');
-  const { isGrouped } = useIsGrouped('stocktake');
-  const { rows, onChangeSortBy, sortBy } = useStocktake.line.rows(isGrouped);
+  const {
```
changes to use the stocktake_lines query

mark-prins commented on client/packages/inventory/src/Stocktake/DetailView/ContentArea/columns.ts:
```diff
@@ -99,6 +99,7 @@ export const useStocktakeColumns = ({
           accessor: ({ rowData }) => {
             return rowData.item?.unitName ?? '';
           },
+          sortable: false,
```
we don't support this in the api, so I've had to remove, along with those below

jmbrunskill replied:
Maybe should raise and issue for resolving these, or at least review is any are needed? How do we identify if this could be needed by customers?

mark-prins replied:
I'm sure that they are useful, and if we asked the stakeholders they'd request for them to stay sortable.. rather than assume, I'll make an issue 🙏

mark-prins commented on client/packages/inventory/src/Stocktake/DetailView/DetailView.tsx:
```diff
@@ -17,31 +17,29 @@ import { Footer } from './Footer';
-  const t = useTranslation('inventory');
-  const { HighlightStyles } = useRowHighlight();
-
+const StocktakeTabs = ({
```
no, it didn't help much. I've kept it in anyway. 
there was re-rendering due to the TableStore being at the top level, so when that updated all the components had to update. this helps a little

mark-prins commented on client/packages/inventory/src/Stocktake/DetailView/DetailView.tsx:
```diff
@@ -53,23 +51,77 @@ export const DetailView: FC = () => {
+};
+
+export const DetailView = () => {
+  const { data: stocktake, isLoading } = useStocktake.document.get();
```
more refactoring to isolate components, and only render those that need to

mark-prins commented on client/packages/inventory/src/Stocktake/DetailView/Toolbar.tsx:
```diff
@@ -12,26 +12,40 @@ import {
-  const t = useTranslation('inventory');
+export const Toolbar = () => {
+  const { isGrouped, toggleIsGrouped } = useIsGrouped('stocktake');
+  const [localIsGrouped, setLocalIsGrouped] = React.useState(isGrouped);
```
this is the change to separate UI update with the wider state change, that impacts the table and slows thigns down.

jmbrunskill reviewed (commented):
Is this correct? e.g. displaying Multiple when there are no locations?
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/e37a855e-9e03-41e4-9860-6f850b13e47d)

There's a double scroll bar...
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/3ccc13a9-ae4e-443d-9a94-36795ece4d0a)

Is expand/collapse all supposed to do something more than this?
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/2d98e85e-aa3b-4105-8849-8f5c3ba1efb0)
vs
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/e731e1dc-0ac0-4be0-bdd8-43b2545f2a0d)

While stocktake page is loading we get a Nothing here message?

![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/6712071f-fb03-4b7f-9edd-b336a2c3b2c3)


Also on android, it's not obvious there's pagination? (At least when used via chrome connection to the server)

Updating to 1.6.01 seemed to ask for re-inititalisation?
![telegram-cloud-photo-size-1-5069353129740446875-y](https://github.com/msupply-foundation/open-msupply/assets/8287373/98eb5c0d-9d45-47ff-aa81-fd58e6eab437)

jmbrunskill reviewed (commented):
Looks like the pagination not showing on the screen does affect the proper android build too...

Stocktakes
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/d4f3d3a7-a609-4502-927f-7c7dcc509be8)

VS

A Stocktake...
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/5b8dfa0f-89af-4a6d-86d7-1a0165f624a9)
(Pagination is hidden behind a scroll bar...

jmbrunskill commented on client/packages/inventory/src/Stocktake/DetailView/ContentArea/columns.ts:
```diff
@@ -282,17 +226,12 @@ export const useStocktakeColumns = ({
         label: 'label.reason',
         accessor: ({ rowData }) => getStocktakeReasons(rowData, t),
-        getSortValue: rowData => getStocktakeReasons(rowData, t),
+        sortable: false,
```
I can imagine sort by reason might be useful?

jmbrunskill commented on client/packages/inventory/src/Stocktake/DetailView/ContentArea/columns.ts:
```diff
@@ -239,25 +201,7 @@ export const useStocktakeColumns = ({
-            );
-          }
-        },
+        sortable: false,
```
Sorting by difference also doesn't seem too important? Although could be handy to see which items appear to have gone missing?

jmbrunskill commented on client/packages/inventory/src/Stocktake/DetailView/ContentArea/columns.ts:
```diff
@@ -208,19 +182,7 @@ export const useStocktakeColumns = ({
-            return row.countedNumberOfPacks ?? '';
-          }
-        },
+        sortable: false,
```
Sorting by number of packs doesn't sound particularly useful...

jmbrunskill commented on client/packages/inventory/src/Stocktake/DetailView/DetailView.tsx:
```diff
@@ -17,31 +17,29 @@ import { Footer } from './Footer';
     [onOpen]
   );
 
-  if (isLoading) return <DetailViewSkeleton hasGroupBy={true} hasHold={true} />;
```
Is removing this why we're seeing the 'No Data' while loading?

mark-prins replied:
It's been moved up a little, rather than removed.
The 'No Data' is something else - I felt that I'd spent long enough on the issue, so didn't spend time debugging that one, will have another look now that the PR is out. it's bothering me.

jmbrunskill commented on client/packages/inventory/src/Stocktake/DetailView/SidePanel.tsx:
```diff
@@ -52,30 +52,30 @@ const AdditionalInfoSection: FC = () => {
     navigator.clipboard
-      .writeText(JSON.stringify(data, null, 4) ?? '')
+      .writeText(JSON.stringify(stocktake, null, 4) ?? '')
       .then(() => success('Copied to clipboard successfully')());
```
I'm getting something weird happening with copy to clipboard on android?


https://github.com/msupply-foundation/open-msupply/assets/8287373/ce27edb1-00bc-45b4-be41-ad52c4eb333d

mark-prins replied:
the function is not supported on android, and I thought the buttons were removed on the android platform.

jmbrunskill replied:
Want me to create an issue?

jmbrunskill commented on server/repository/src/db_diesel/stocktake_line.rs:
```diff
@@ -175,3 +196,25 @@ fn create_filtered_query(filter: Option<StocktakeLineFilter>) -> BoxedStocktakeL
+            item_filter.is_visible = Some(true);
+            let items = ItemRepository::new(connection)
+                .query_by_filter(item_filter, Some(store_id))
+                .unwrap();
```
Is this unwrap safe? If so can you comment why?

mark-prins replied:
I copied it from the stock_line filter, and figured that if it's good enough for them, then it's good enough for me.

jmbrunskill commented on server/service/src/stocktake/validate.rs:
```diff
@@ -18,10 +18,12 @@ pub fn check_stocktake_not_finalised(status: &StocktakeStatus) -> bool {
-    ))?;
+    let result = StocktakeLineRepository::new(connection).count(
+        Some(StocktakeLineFilter::new().stocktake_id(EqualFilter::equal_to(stocktake_line_id))),
+        Some(store_id.to_string()),
```
Is is safer to have this a `None` just incase the same id could exist at a different store?

mark-prins replied:
the store_id is only for use in the item code or name filter, which isn't used in this call - so passing in store_id or None would have the same effect.. though could be confusing to people reading the code

jmbrunskill reviewed (approved):
Generally I'm good with this change, but would be good to have another review as I'm not too familiar with current issues.

Not seeing the pagination on the stocktake screen by default feels a bit broken to me, but I'm happy if its a better experience than the current slowness?
Not sure if other screens need to scroll to see pagination already and maybe is familiar to the users?

The other area I'm not 100% confident on is adding the store_id to the queries. Is it possible that affects some other parts of the app?

mark-prins commented:
Thanks James for the review! 
I agree - the No Data message is bothering me to.. that shows in the space between when the query returns, and the table gets the data. that's probably the bit I should look at speeding up, so will have another go. I had thought that I'd spent too long on the issue already and wanted to get something out there, so didn't spend too long on that part.

The sorting implementation: I decided that was taking the code changes too far and that implementing the sort is a separate issue. And the pagination bothered me too, but same approach: spent long enough, and thought that was a small price to pay for having the stocktake load. Ideally it should be static below the table and always visible, as well as fitting in horizontally. the latter part I've already tried and failed to make it work nicely.

jmbrunskill commented:
> . Ideally it should be static below the table and always visible, as well as fitting in horizontally. the latter part I've already tried and failed to make it work nicely.

Maybe create a follow up issue for Triage?

mark-prins commented:
thanks again @jmbrunskill - good motivation there.
I've done this:
- fixed the pagination at the bottom, so that it is always visible
- added a loader for the lines query
- removed the `unwrap` and commented
- added functionality to handle the pagination when the modal is open and you click `Ok & Next`

In that last case - if you are at the end of the current page of results, and there are more pages, the `Ok & Next` will fetch the next page and set the currently edited line to be the first of the next set of results. When you reach the very end of the results & pages, `Ok & Next` is disabled. It's more intuitive like this 🙏

clemens-msupply commented on client/packages/inventory/src/Stocktake/DetailView/modal/StocktakeLineEdit/StocktakeLineEdit.tsx:
```diff
@@ -65,9 +71,26 @@ export const StocktakeLineEdit: FC<StocktakeLineEditProps> = ({
+        // so that we can correctly set the current item when the
+        // lines query returns
+        updatePaginationQuery(page + 1);
+        setCurrentItem(null);
```
Updating the pagination will trigger fetching the next item using `useStocktake.line.rows()`, right?

mark-prins replied:
the current item is in local state, so the updated list of items from `useStocktake.line.rows()` won't help if that's what you mean. Yes, the pagination will call the lines query and fetch items in `useStocktake.line.rows()`

clemens-msupply commented on client/packages/inventory/src/Stocktake/DetailView/modal/StocktakeLineEdit/hooks/useStocktakeLineEdit.ts:
```diff
@@ -25,7 +25,9 @@ export const useStocktakeLineEdit = (
   const { id } = useStocktake.document.fields('id');
-  const nextItem = useNextItem(item?.id);
+  const { items } = useStocktake.line.rows();
+  const filtereredItems = items.filter(item => item.item?.id === item?.id);
```
Spelling...

clemens-msupply replied:
There can be multiple stocktage lines for a single item, right?

mark-prins replied:
this is ironic, I think. picking up a spelling error and making one in the following comment!
there can be multiple stocktake lines for an item yep.

clemens-msupply commented on client/packages/common/src/ui/components/navigation/Tabs/DetailTabs.tsx:
```diff
@@ -45,6 +50,8 @@ export const DetailTabs: FC<DetailTabsProps> = ({
   useEffect(() => {
     handleResize();
+    // depending on handleResize causes too many renders
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
I guess you can move the definition of handleResize out of the component to fix this?

mark-prins replied:
yes! I thought I had. thanks, will do

clemens-msupply commented on client/packages/common/src/ui/components/navigation/Tabs/DetailTabs.tsx:
```diff
@@ -89,6 +91,8 @@ export const DetailTabs: FC<DetailTabsProps> = ({
       });
     }
+    // isValidTab is recreated on every render, so it's safe to ignore it
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Should `isValidTab` be wrapped into a useCallback() instead?

mark-prins replied:
yes, that's more correct. have done thanks.

clemens-msupply reviewed (approved):
Only had a look at the code yet and trust James for his testing. Looks good to me. A couple of nits that should maybe be addressed in a follow up issue, e.g. use sub query to fetch items.

mark-prins commented on client/packages/inventory/src/Stocktake/DetailView/Toolbar.tsx:
```diff
@@ -12,26 +12,40 @@ import {
+    // when the render of the dependent component is slow
+    // separate the render of the switch change from the wider state change
+    // otherwise the switch doesn't render until the slow component completes
+    setTimeout(toggleIsGrouped, 100);
```
I think this speaks to the underlying issue - we have too many react components which is making the rendering slow. This is a little hacky - an alternative is for the UI to be unresponsive, so I'll take the hacky for now.

mark-prins commented on server/repository/src/db_diesel/stocktake_line.rs:
```diff
@@ -89,25 +92,43 @@ impl<'a> StocktakeLineRepository<'a> {
         pagination: Pagination,
         filter: Option<StocktakeLineFilter>,
         sort: Option<StocktakeLineSort>,
+        store_id: Option<String>,
```
cool - thanks. have done.

mark-prins commented:
Thanks for reviews! @clemens-msupply I've created a follow up issue (#2939) to address the query consolidation concerns. I don't want to rush the change in here.
